### PR TITLE
Add SMS/MMS Messagebird Channel

### PR DESCRIPTION
**Pull request by tybritten**
Related courier PR: 
https://github.com/nyaruka/courier/pull/608

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -56,6 +56,7 @@ ffmpeg-python = "^0.2.0"
 slack-sdk = "3.17.0"
 django-formtools = "^2.4.1"
 psycopg = "^3.1.9"
+messagebird = "^2.1.0"
```
Comment (norkans7) [inline pyproject.toml:59]: It seems we do not need to install this? do we use that?

File: temba/channels/types/messagebird/views.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+
+class ClaimView(ClaimViewMixin, SmartFormView):
+    class Form(ClaimViewMixin.Form):
+        phone_num = forms.CharField(
+            required=True,
+            label=_("Originating Phone number"),
+            help_text=_("The phone number being added"),
+        )
+
+        def clean_phone_number(self):
```
Comment (norkans7) [inline temba/channels/types/messagebird/views.py:61]: move this below the fields

File: temba/channels/types/messagebird/views.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+    "ZA",  # South Africa  # Beta
+}
+
+COUNTRY_CHOICES = countries.choices(SUPPORTED_COUNTRIES)
+CALLING_CODES = countries.calling_codes(SUPPORTED_COUNTRIES)
+
+
+class ClaimView(ClaimViewMixin, SmartFormView):
+    class Form(ClaimViewMixin.Form):
+        phone_num = forms.CharField(
```
Comment (norkans7) [inline temba/channels/types/messagebird/views.py:55]: I guess this should be `phone_number`, probably `number` is just enough

File: temba/channels/types/messagebird/type.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+    name = "MessageBird"
+
+    claim_blurb = _("Connect your approved %(link)s channel") % {
+        "link": '<a target="_blank" href="https://www.messagebird.com/">Messagebird</a>'
+    }
+    claim_view = ClaimView
+
+    schemes = [URN.TEL_SCHEME]
+    free_sending = True
```
Comment (norkans7) [inline temba/channels/types/messagebird/type.py:27]: I guess this is not `free_sending` as that uses SMS

File: temba/channels/types/messagebird/type.py [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
```
Comment (norkans7) [inline temba/channels/types/messagebird/type.py:28]: set the `available_timezones` with a list of timezones for the supported countries

Comment (norkans7) [reply temba/channels/types/messagebird/type.py:28]: That way the channel is shown where it is supported

File: temba/channels/types/messagebird/views.py [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+
+            # short code should not be formatted
+            if len(phone) <= 6:
+                return phone
+
+            phone = phonenumbers.parse(phone, self.cleaned_data["country"])
+            return phonenumbers.format_number(phone, phonenumbers.PhoneNumberFormat.E164)
+
+        country = forms.ChoiceField(choices=COUNTRY_CHOICES, widget=SelectWidget(attrs={"searchable": True}))
+        signing_key = forms.CharField(
```
Comment (norkans7) [inline temba/channels/types/messagebird/views.py:72]: field could be called `secret`

File: temba/channels/types/messagebird/type.py [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+        "receivedSMS messages to the url below either with a flow or by registering the webhook with them"
+        "Shortcodes don't work with flows and require a webhook."
+    )
+
+    configuration_urls = (
+        dict(
+            label=_("Receive URL"),
+            url="https://{{ channel.callback_domain }}{% url 'courier.mbd' channel.uuid 'receive'%}",
+            description=_("Webhook address for inbmound messages to this address."),
+        ),
```
Comment (norkans7) [inline temba/channels/types/messagebird/type.py:40]: Add Status URL dict here too

File: temba/channels/types/messagebird/type.py [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+        "Shortcodes don't work with flows and require a webhook."
+        "Configure the status url under Developer Settings to receive status updates for your messages."
+    )
+
+    configuration_urls = (
+        dict(
+            label=_("Receive URL"),
+            url="https://{{ channel.callback_domain }}{% url 'courier.mbd' channel.uuid 'receive'%}",
+            description=_("Webhook address for inbmound messages to this address."),
+            label=_("Status URL"),
```
[the remark this answers is not in the document]
Comment (tybritten) [reply temba/channels/types/messagebird/type.py:43]: ahhh ok

File: temba/channels/types/messagebird/tests.py [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+
+    def test_claim(self):
+        url = reverse("channels.types.messagebird.claim")
+
+        self.login(self.admin)
+
+        # check that claim page URL appears on claim list page
+        response = self.client.get(reverse("channels.channel_claim"))
+        self.assertContains(response, url)
+
```
Comment (norkans7) [inline temba/channels/types/messagebird/tests.py:35]: Add tests to make sure the channel type is only available for beta group only here

File: temba/channels/types/messagebird/tests.py [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+        self.channel = Channel.create(
+            self.org,
+            self.user,
+            "US",
+            "MBD",
+            name="Messagebird: 12345",
+            address="12345",
+            role="SR",
+            schemes=["freshchat"],
```
[the remark this answers is not in the document]
Comment (norkans7) [reply temba/channels/types/messagebird/tests.py:20]: still not solved, and that breaks the tests @tybritten double check the schemes here

File: temba/channels/types/messagebird/tests.py (hunk 9 above)
Comment (tybritten) [reply temba/channels/types/messagebird/tests.py]: I added this test and it doesn't seem to be working right?
```        # beta- should not see the general channel claim page
        response = self.client.get(reverse("channels.channel_claim"))
        self.assertNotContains(response, url)

        self.make_beta(self.admin)

        # check that claim page URL appears on claim list page
        response = self.client.get(reverse("channels.channel_claim"))
        self.assertContains(response, url)```

**Review by norkans7 — COMMENTED** [review]
Tests still failing, that might be related to the timezone of the org in the test, make sure that is changed one of the supported timezone of the channel type.

**Comment by tybritten** [comment]
Ahhh.. I had capitalization mismatch in my test.

**Comment by tybritten** [comment]
Ok I have this fixed. It was a couple overlapping issues.

**Review by norkans7 — APPROVED** [review]
Looks good to me

**Pull request merged**